Haptics/app/src/main/java/com/example/android/haptics/samples/ui/views/HomeViewModel.kt

```diff
@@ -0,0 +1,257 @@
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+                // https://developer.android.com/reference/android/os/VibrationEffect#createPredefined(int)
+                vibrator.vibrate(VibrationEffect.createPredefined(effect))
+            }
+        } else if (category == HapticFeedbackCategoryType.HAPTIC_CONSTANTS) {
+            // https://developer.android.com/reference/android/view/View#performHapticFeedback(int)
+            view.performHapticFeedback(effect)
+        } else if (category == HapticFeedbackCategoryType.PREDEFINED_PRIMITIVES) {
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+                // https://developer.android.com/reference/android/os/VibrationEffect.Composition
+                vibrator.vibrate(
+                    VibrationEffect.startComposition()
+                        .addPrimitive(effect, 0.8f)
+                        .compose()
+                )
+            }
+        } else {
+            throw RuntimeException("No corresponding action for HapticFeedbackCategoryType.$category.")
+        }
+    }
+
+    fun onSnackbarMessage(message: String) {
+        viewModelScope.launch {
+            scaffoldState.snackbarHostState.showSnackbar(message)
+        }
+    }
+
+    /**
+     * Factory for HomeViewModel that can provide vibrator dependency and construct
+     * HomeViewModelState using Application context.
+     */
+    companion object {
+        fun provideFactory(
+            application: Application,
+            scaffoldState: ScaffoldState,
+            scrollState: ScrollState,
+        ): ViewModelProvider.Factory = object : ViewModelProvider.Factory {
+            @Suppress("UNCHECKED_CAST")
+            override fun <T : ViewModel> create(modelClass: Class<T>): T {
+                val androidRorLater = Build.VERSION.SDK_INT >= Build.VERSION_CODES.R // API Level 30.
+                val vibrator = application.getSystemService(Vibrator::class.java)
+
+                // This state object is responsible for providing all the information necessary
+                // for the UI to build the home page of buttons, separated into categories.
+                val viewModelState = HomeViewModelState(
+                    hapticFeedbackCategories = listOf(
+                        // List of predefined effects.
+                        HapticFeedbackCategory(
+                            application.getString(R.string.home_screen_predefined_effects),
+                            categoryType = HapticFeedbackCategoryType.PREDEFINED_EFFECTS,
+                            buttons = listOf(
+                                HapticFeedbackButton(
+                                    label = application.getString(R.string.home_screen_tick),
+                                    worksOnUserDevice = isEffectSupported(
+                                        vibrator,
+                                        VibrationEffect.EFFECT_TICK
```

Comment:
Does this work in sdk < Q? The constant was only introduced then, so I think that attempting to use it before checking will crash. Same for all constants below. I think this needs a different kind of helper method to handle SDK versions.